server/storage/bootstrap.go

```diff
@@ -0,0 +1,335 @@
+		}
+		if cfg.UnsafeNoFsync {
+			w.SetUnsafeNoFsync()
+		}
+		wmetadata, st, ents, err := w.ReadAll()
+		if err != nil {
+			w.Close()
+			// we can only repair ErrUnexpectedEOF and we never repair twice.
+			if repaired || err != io.ErrUnexpectedEOF {
+				cfg.Logger.Fatal("failed to read WAL, cannot Be repaired", zap.Error(err))
+			}
+			if !wal.Repair(cfg.Logger, cfg.WALDir()) {
+				cfg.Logger.Fatal("failed to repair WAL", zap.Error(err))
+			} else {
+				cfg.Logger.Info("repaired WAL", zap.Error(err))
+				repaired = true
+			}
+			continue
+		}
+		var m etcdserverpb.Metadata
+		pbutil.MustUnmarshal(&m, wmetadata)
+		id := types.ID(m.NodeID)
+		cid := types.ID(m.ClusterID)
+		meta := &Metadata{ClusterID: cid, NodeID: id}
+
+		wal := &BootstrappedWAL{
+			lg:       cfg.Logger,
+			HaveWAL:  true,
+			Wal:      w,
+			State:    &st,
+			Entries:  ents,
+			Snapshot: snapshot,
+			Meta:     meta,
+		}
+
+		if cfg.ForceNewCluster {
+			// discard the previously uncommitted entries
+			wal.Entries = wal.CommitedEntries()
+			entries := wal.ConfigChangeEntries(meta)
+			// force commit config change entries
+			wal.AppendAndCommitEntries(entries)
+			cfg.Logger.Info(
+				"forcing restart member",
+				zap.String("cluster-id", meta.ClusterID.String()),
+				zap.String("local-member-id", meta.NodeID.String()),
+				zap.Uint64("commit-index", wal.State.Commit),
+			)
+		} else {
+			cfg.Logger.Info(
+				"restarting local member",
+				zap.String("cluster-id", meta.ClusterID.String()),
+				zap.String("local-member-id", meta.NodeID.String()),
+				zap.Uint64("commit-index", wal.State.Commit),
+			)
+		}
+		return wal
+	}
+}
+
+func bootstrapNewWAL(cfg config.ServerConfig, meta *Metadata) *BootstrappedWAL {
+	metadata := pbutil.MustMarshal(
+		&etcdserverpb.Metadata{
+			NodeID:    uint64(meta.NodeID),
+			ClusterID: uint64(meta.ClusterID),
+		},
+	)
+	w, err := wal.Create(cfg.Logger, cfg.WALDir(), metadata)
+	if err != nil {
+		cfg.Logger.Panic("failed to create WAL", zap.Error(err))
+	}
+	if cfg.UnsafeNoFsync {
+		w.SetUnsafeNoFsync()
+	}
+	return &BootstrappedWAL{
+		lg:  cfg.Logger,
+		Wal: w,
+	}
+}
+
+func (wal *BootstrappedWAL) MemoryStorage() *raft.MemoryStorage {
+	s := raft.NewMemoryStorage()
+	if wal.Snapshot != nil {
+		s.ApplySnapshot(*wal.Snapshot)
+	}
+	if wal.State != nil {
+		s.SetHardState(*wal.State)
+	}
+	if len(wal.Entries) != 0 {
+		s.Append(wal.Entries)
+	}
+	return s
+}
+
+func (wal *BootstrappedWAL) CommitedEntries() []raftpb.Entry {
+	for i, ent := range wal.Entries {
+		if ent.Index > wal.State.Commit {
+			wal.lg.Info(
+				"discarding uncommitted WAL entries",
+				zap.Uint64("entry-index", ent.Index),
+				zap.Uint64("commit-index-from-Wal", wal.State.Commit),
+				zap.Int("number-of-discarded-entries", len(wal.Entries)-i),
+			)
+			return wal.Entries[:i]
+		}
+	}
+	return wal.Entries
+}
+
+func (wal *BootstrappedWAL) ConfigChangeEntries(meta *Metadata) []raftpb.Entry {
```

Comment:
NewConfigChangeEntries (without this it sounds like reading them from wal)
or GenerateConfigChangeEntries

Reply:
why do you pass 'meta' directly instead of reuse wal.meta ?

Reply:
Done